File: kt/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/text/SourceText.kt
package dev.phytolizer.minsk.analysis.text

class SourceText private constructor(private val _text: String) {
    val lines = parseLines(this, _text)
    val length = _text.length

    companion object {
        fun from(text: String): SourceText {
            return SourceText(text)
        }

        private fun parseLines(sourceText: SourceText, text: String): List<TextLine> {
            val result = mutableListOf<TextLine>()

            var position = 0
            var lineStart = 0

            while (position < text.length) {
                val lineBreakWidth = getLineBreakWidth(text, position)

                if (lineBreakWidth == 0) {
                    position += 1
                } else {
                    addLine(result, sourceText, position, lineStart, lineBreakWidth)

                    position += lineBreakWidth
                    lineStart = position
                }
            }

            addLine(result, sourceText, position, lineStart, 0)

            return result
        }

        private fun addLine(
            lines: MutableList<TextLine>,
            sourceText: SourceText,
            position: Int,
            lineStart: Int,
            lineBreakWidth: Int,
        ) {
            val lineLength = position - lineStart
            val lineLengthIncludingLineBreak = lineLength + lineBreakWidth
            val line = TextLine(sourceText, lineStart, lineLength, lineLengthIncludingLineBreak)
            lines.add(line)
        }

        private fun getLineBreakWidth(text: String, position: Int): Int {
            val char = text[position]
            val look = if (position + 1 >= text.length) {
                '\u0000'
            } else {
                text[position + 1]
            }

            return if (char == '\r' && look == '\n') {
                2
            } else if (char == '\r' || char == '\n') {
                1
            } else {
                0
            }
        }
    }

    fun lineIndex(position: Int): Int {
        var lower = 0
        var upper = lines.size - 1

        while (lower <= upper) {
            val index = lower + (upper - lower) / 2
            val start = lines[index].start

            if (position == start) {
                return index
            }

            if (start > position) {
                upper = index - 1
            } else {
                lower = index + 1
            }
        }

        return lower - 1
    }

    operator fun get(index: Int) = _text[index]

    override fun toString() = _text
    fun toString(range: IntRange) = _text.substring(range)
    fun toString(span: TextSpan) = _text.substring(span.range)
}

File: cs/Minsk/CodeAnalysis/Text/SourceText.cs
using System.Collections.Immutable;

namespace Minsk.CodeAnalysis.Text;

public sealed class SourceText
{
    private readonly string _text;

    private SourceText(string text)
    {
        _text = text;
        Lines = ParseLines(this, text);
    }

    public ImmutableArray<TextLine> Lines { get; }
    public char this[int index] => _text[index];
    public int Length => _text.Length;

    public int GetLineIndex(int position)
    {
        var lower = 0;
        var upper = Lines.Length - 1;

        while (lower <= upper)
        {
            var index = lower + (upper - lower) / 2;
            var start = Lines[index].Start;

            if (position == start)
            {
                return index;
            }

            if (start > position)
            {
                upper = index - 1;
            }
            else
            {
                lower = index + 1;
            }
        }

        return lower - 1;
    }

    private static ImmutableArray<TextLine> ParseLines(SourceText sourceText, string text)
    {
        var result = ImmutableArray.CreateBuilder<TextLine>();
        var position = 0;
        var lineStart = 0;

        while (position < text.Length)
        {
            var lineBreakWidth = GetLineBreakWidth(text, position);
            if (lineBreakWidth == 0)
            {
                position++;
            }
            else
            {
                AddLine(result, sourceText, position, lineStart, lineBreakWidth);

                position += lineBreakWidth;
                lineStart = position;
            }
        }

        AddLine(result, sourceText, position, lineStart, 0);

        return result.ToImmutable();
    }

    private static void AddLine(
        ImmutableArray<TextLine>.Builder result,
        SourceText sourceText,
        int position,
        int lineStart,
        int lineBreakWidth
    )
    {
        var lineLength = position - lineStart;
        var lineLengthIncludingLineBreak = lineLength + lineBreakWidth;
        var line = new TextLine(sourceText, lineStart, lineLength, lineLengthIncludingLineBreak);
        result.Add(line);
    }

    private static int GetLineBreakWidth(string text, int position)
    {
        var c = text[position];
        var l = position + 1 >= text.Length ? '\0' : text[position + 1];

        return c switch
        {
            '\r' when l == '\n' => 2,
            '\r' or '\n' => 1,
            _ => 0
        };
    }

    public static SourceText From(string text)
    {
        return new SourceText(text);
    }

    public override string ToString()
    {
        return _text;
    }

    public string ToString(int start, int length)
    {
        return _text.Substring(start, length);
    }

    public string ToString(TextSpan span)
    {
        var (start, length) = span;
        return ToString(start, length);
    }
}
